File: 0097-interleaving-string/0097-interleaving-string.ts
const helper = (
  s1: string,
  s2: string,
  s3: string,
  i: number,
  j: number,
  memo: Map<string, boolean>
): boolean => {
  const key = `${i}|${j}`;

  if (memo.has(key)) return memo.get(key)!;

  const k = i + j;

  if (k === s3.length) return true;

  let result = false;

  if (i < s1.length && s1[i] === s3[k]) {
    result = helper(s1, s2, s3, i + 1, j, memo);
  }

  if (!result && j < s2.length && s2[j] === s3[k]) {
    result = helper(s1, s2, s3, i, j + 1, memo);
  }

  memo.set(key, result);
  return result;
};

function isInterleave(s1: string, s2: string, s3: string): boolean {
  if (s1.length + s2.length !== s3.length) return false;

  return helper(s1, s2, s3, 0, 0, new Map());
}

File: 0097-interleaving-string/0097-interleaving-string.java
class Solution {

    boolean helper(String s1, String s2, String s3, int i, int j, int k, Boolean[][][] dp) {

        int m = s1.length();
        int n = s2.length();
        int o = s3.length();

        if (i == m && j == n && k == o)
            return true;
        if (dp[i][j][k] != null)
            return dp[i][j][k];
        if (i < m && j < n && k < o && s3.charAt(k) == s1.charAt(i) && s3.charAt(k) == s2.charAt(j)) {
            return dp[i][j][k] = helper(s1, s2, s3, i + 1, j, k + 1, dp) || helper(s1, s2, s3, i, j + 1, k + 1, dp);
        } else if (k < o && i < m && s3.charAt(k) == s1.charAt(i)) {
            return dp[i][j][k] = helper(s1, s2, s3, i + 1, j, k + 1, dp);
        } else if (k < o && j < n && s3.charAt(k) == s2.charAt(j)) {
            return dp[i][j][k] = helper(s1, s2, s3, i, j + 1, k + 1, dp);
        } else
            return dp[i][j][k] = false;
    }

    public boolean isInterleave(String s1, String s2, String s3) {
        int m = s1.length();
        int n = s2.length();
        int o = s3.length();
        Boolean[][][] dp = new Boolean[m+1][n+1][o+1];
        return helper(s1, s2, s3, 0, 0, 0, dp);
    }
}